python/cudf/cudf/core/buffer/exposure_tracked_buffer.py

```diff
@@ -0,0 +1,311 @@
+# Copyright (c) 2020-2023, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+import weakref
+from typing import (
+    Any,
+    Container,
+    Literal,
+    Mapping,
+    Optional,
+    Type,
+    TypeVar,
+    cast,
+)
+
+from typing_extensions import Self
+
+import cudf
+from cudf.core.buffer.buffer import Buffer, get_ptr_and_size
+from cudf.utils.string import format_bytes
+
+T = TypeVar("T", bound="ExposureTrackedBuffer")
+
+
+def get_owner(data, klass: Type[T]) -> Optional[T]:
+    """Get the owner of `data`, if any exist
+
+    Search through the stack of data owners in order to find an
+    owner of type `klass` (not subclasses).
+
+    Parameters
+    ----------
+    data
+        The data object
+
+    Return
+    ------
+    klass or None
+        The owner of `data` if `klass` or None.
+    """
+
+    if type(data) is klass:
+        return data
+    if hasattr(data, "owner"):
+        return get_owner(data.owner, klass)
+    return None
+
+
+def as_exposure_tracked_buffer(
+    data, exposed: bool, subclass: Optional[Type[T]] = None
+) -> BufferSlice:
+    """Factory function to wrap `data` in a slice of an exposure tracked buffer
+
+    If `subclass` is None, a new ExposureTrackedBuffer that points to the
+    memory of `data` is created and a BufferSlice that points to all of the
+    new ExposureTrackedBuffer is returned.
+
+    If `subclass` is not None, a new `subclass` is created instead. Still,
+    a BufferSlice that points to all of the new `subclass` is returned
+
+    It is illegal for an exposure tracked buffer to own another exposure
+    tracked buffer. When representing the same memory, we should have a single
+    exposure tracked buffer and multiple buffer slices.
+
+    Developer Notes
+    ---------------
+    This function always returns slices thus all buffers in cudf will use
+    `BufferSlice` when copy-on-write is enabled. The slices implements
+    copy-on-write by trigging deep copies when write access is detected
+    and multiple slices points to the same exposure tracked buffer.
+
+    Parameters
+    ----------
+    data : buffer-like or array-like
+        A buffer-like or array-like object that represent C-contiguous memory.
+    exposed
+        Mark the buffer as permanently exposed.
+    subclass
+        If not None, a subclass of ExposureTrackedBuffer to wrap `data`.
+
+    Return
+    ------
+    BufferSlice
+        A buffer slice that points to a ExposureTrackedBuffer (or `subclass`),
+        which in turn wraps `data`.
+    """
+
+    if not hasattr(data, "__cuda_array_interface__"):
+        if exposed:
+            raise ValueError("cannot created exposed host memory")
+        return cast(
+            BufferSlice, ExposureTrackedBuffer._from_host_memory(data)[:]
```

Comment:
Is the eventual plan for Buffer.__getitem__ to return a `BufferSlice`? If not, it might be cleaner to override the method in `ExposureTrackedBuffer`. I know the whole point of the `_getitem`/`__getitem__` split is to help share some functionality, but the typing confusion here indicates that there are potentially incorrect types that could result from that approach (obviously we can coerce the code into behaving correctly, but it makes it much harder to write intrinsically type-safe code if the type annotations aren't sufficiently valid).

Reply:
I agree, this is a very valid point!

I think the clean design is to make cudf always work on `BufferSlice` even when COW and spilling is disabled. Then we get a clean class hierarchy: 
- COW & Spilling disable:
`BufferSlice -> Buffer`
- COW enabled:
`BufferSlice -> ExposureTrackedBuffer -> Buffer`
- Spilling enabled (when is has been unified with COW in a follow-up PR):
`BufferSlice -> SpillableBuffer -> ExposureTrackedBuffer -> Buffer`

The downside is that this approach is a bit more intrusive in the default case where COW and spilling is disabled. I think it is worth it but what do you guys think?

cc. @wence- @shwina

Reply:
I agree that it is probably worth it. I would stage that as work to be done after the COW and spilling unification is complete and we can reevaluate in the context of a cleaner architecture.